### ISSUE TITLE
FR: CSR (Cloud Source Repositories) EOL June 2024 - replace the default path for version control system and CICD tool

### ISSUE THREAD

**Issue by fmichaelobrien** [terraform-google-modules/terraform-example-foundation#1249]
### TL;DR

CSR is undergoing deprecation in favour of SSM
- https://cloud.google.com/source-repositories/docs/authentication
- https://cloud.google.com/secure-source-manager/docs/overview

## Impact to CICD
- will affect the CB/CSR default option in the TEF LZ - primarily 0-bootstrap starting with the following gcloud clone
- https://github.com/terraform-google-modules/terraform-example-foundation/blob/master/0-bootstrap/scripts/push-to-repo.sh#L32
- CB/Cloud Build will continue to be used for the pipeline
- will not affect incoming ADO option in #1205 
- shadow https://github.com/GoogleCloudPlatform/pbmm-on-gcp-onboarding/issues/439
- reference SSH authentication option over gcloud API clone in https://github.com/GoogleCloudPlatform/pbmm-on-gcp-onboarding/issues/431

```
Cloud Source Repositories is scheduled for end of sale on June 17, 2024. Starting June 17, 2024, if your organization hasn't previously used Cloud Source Repositories, you cannot enable the API or use Cloud Source Repositories. New projects not connected to an organization can’t enable the Cloud Source Repositories API after June 17, 2024. Customers who have already enabled the API prior to this date will not be affected and can continue to use Cloud Source Repositories.
```
<img width="1630" alt="Screenshot 2024-05-22 at 7 40 08 AM" src="https://github.com/terraform-google-modules/terraform-example-foundation/assets/94715080/7b00703d-38c8-47e6-ac36-7551520326e9">

<img width="1384" alt="Screenshot 2024-05-22 at 7 39 32 AM" src="https://github.com/terraform-google-modules/terraform-example-foundation/assets/94715080/68a046f6-b5e5-4455-bfbb-030371384a08">

### Terraform Resources

```markdown
1.3.10
```

### Detailed design

```markdown
Work is in progress in also bringing in ADO (Azure DevOps) as a CI/CD option - as it is the default repository/pipeline tool for 80% of CA PubSec clients

https://github.com/terraform-google-modules/terraform-example-foundation/issues/1205
```

### Additional information

fmichaelobrien will look into the SSM addition unless this work is already assigned in the roadmap

**Comment by obriensystems** [comment]
module references
- https://github.com/terraform-google-modules/terraform-google-bootstrap/blob/master/modules/tf_cloudbuild_source/main.tf

**Comment by eeaton** [comment]
Thanks for raising this issue. This is identified on our internal roadmap as work to address in the next major round of updates for v5 (sometime in H2 this year). The answer may or may not be SSM, due to the limitation that SSM is currently an invitation only service.

**Comment by sleighton2022** [comment]
Closing this issue, and as eeaton noted, we will address this as part of our H2 roadmap.

**Comment by eeaton** [comment]
One additional aspect to address when we work on this:

It was identified in #1273 that running the docker tests documented in CONTRIBUTING.md also have a dependency on Cloud Build & CSR, so this will fail in any new organizations that haven't previously used CSR. It's not yet clear to me whether we can unpick this locally, or it's an upstream issue with the CFT test framework.

**Comment by Marrangas** [comment]
As noted SSM is invitation only. Most vcs can be  currently integrated with cloud buls 2nd generation repos.

> Thanks for raising this issue. This is identified on our internal roadmap as work to address in the next major round of updates for v5 (sometime in H2 this year). The answer may or may not be SSM, due to the limitation that SSM is currently an invitation only service.

Fully automating that cloud build repository connection can be a little fussy. It is possible to expose a veriable `vsc_cb2nd_enabled = false` to wait until that connection is made manually and then change it, to be able to bootstrap with the current tools and still create the pipelines in cloud build.

**Comment by eeaton** [comment]
Tracking additional feedback from [this thread](https://github.com/terraform-google-modules/terraform-example-foundation/pull/1074#issuecomment-2245162755) that directions for [deploying locally in 0-boostrap](https://github.com/terraform-google-modules/terraform-example-foundation/blob/master/0-bootstrap/README.md#deploying-with-cloud-build) are unclear.

There was an assumption before that the Deploy with Cloud Build directions in 0-bootstrap were the same for deploying locally, however 

- that's not clear in the text (deploy locally directions never actually say that)
- the deploy cloud build directions have some dependencies on CSR, so don't work for deploying locally without some understanding of which parts to keep and which to ignore

We'll need to address this as part of reconfiguring the default deployment path and rewriting directions.

**Comment by caetano-colin** [comment]
Issue is being tracked on the following PR:

https://github.com/terraform-google-modules/terraform-example-foundation/pull/1329

**Comment by obriensystems** [comment]
older gcp organizations up to for example 2022, 2023 still have CSR capabilities
for example
https://github.com/ObrienlabsDev/canary-java-springboot/blob/main/cloudbuild.yaml
